### PR TITLE
SNOW-1437407: [Local Testing] Fix convert_timezone argument order

### PR DESCRIPTION
**Pull request by sfc-gh-jrose**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1437407

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   The original mock implementation for convert_timezone assumed that args would be passed in the same order as they are provided to the non-mocked version. Upon further inspection I now see that the args are passed with sql semantics instead. I've updated the mock function to set local variable based on the sql semantics instead and added a test to show that it can be called with all three args.

I've also refactored most of the logic in the function to use pandas.DataFrame.apply instead of converting everything into a list and using an iterator.

### CODE REVIEW

File: src/snowflake/snowpark/mock/_functions.py [hunk 1]
```diff
@@ -1700,42 +1700,53 @@ def mock_initcap(values: ColumnEmulator, delimiters: ColumnEmulator):
...
+            # Using dateutil because it uses iana timezones while pytz would use Olson tzdb.
+            source_time = source_time.replace(tzinfo=dateutil.tz.gettz(source_timezone))
+
+        if source_time.tzinfo is None:
+            source_time = LocalTimezone.replace_tz(source_time)
+
+        result = source_time.astimezone(dateutil.tz.gettz(target_timezone))
+
+        if return_type == TimestampTimeZone.NTZ:
+            result = result.replace(tzinfo=None)
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_functions.py:1745]: this logic seems new, is this also fixing a bug that previously we don't remove tzinfo TimestampNTZ type?

File: tests/integ/scala/test_function_suite.py [hunk 2]
```diff
@@ -3835,6 +3835,52 @@ def test_convert_timezone(session, local_testing_mode):
...
+def test_convert_timezone_neg(session):
+    df = TestData.datetime_primitives1(session)
+    with pytest.raises(SnowparkSQLException):
+        df.select(
+            convert_timezone(lit("UTC"), "timestamp_tz", lit("US/Eastern"))
+        ).collect()
+
+
+def test_convert_timezone_nulls(session):
+    null_df = session.create_dataframe([[None]]).to_df("timestamp")
```
Comment (sfc-gh-aling) [inline tests/integ/scala/test_function_suite.py:3847]: dumb question, do we have test for non-None two args input for convert_timezone?

Comment (sfc-gh-jrose) [reply tests/integ/scala/test_function_suite.py:3847]: Yes, but I've left that one unchanged.

File: src/snowflake/snowpark/mock/_functions.py (hunk 1 above)
Comment (sfc-gh-jrose) [reply src/snowflake/snowpark/mock/_functions.py:1745]: That is correct.

File: src/snowflake/snowpark/mock/_functions.py [hunk 3]
```diff
@@ -1700,42 +1700,53 @@ def mock_initcap(values: ColumnEmulator, delimiters: ColumnEmulator):
 
 @patch("convert_timezone")
 def mock_convert_timezone(
-    target_timezone: ColumnEmulator,
-    source_time: ColumnEmulator,
-    source_timezone: Optional[ColumnEmulator] = None,
+    *args: ColumnEmulator,
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:1703]: Question: what's the motivation behind this change?

Comment (sfc-gh-jrose) [reply src/snowflake/snowpark/mock/_functions.py:1703]: Excerpt from functions.py:
```
    if source_timezone is None:
        return call_builtin("convert_timezone", target_tz, source_time_to_convert)
    return call_builtin(
        "convert_timezone", source_tz, target_tz, source_time_to_convert
    )
```
When convert_timezone is called with 2 parameters they are in a certain order, but when called with 3 the additional parameter is added to the front instead. In practice it seems that the mocked version always gets three parameters, but the ordering is different so I don't think it makes as much sense to name those parameters.

**Review by sfc-gh-stan — APPROVED** [review]

**Review by sfc-gh-aling — APPROVED** [review]
